Sources/MongoSwift/MongoClient.swift

```diff
@@ -2,7 +2,7 @@ import Foundation
 import mongoc
 
 /// Options to use when creating a `MongoClient`.
-public struct ClientOptions {
+public struct ClientOptions: CodingStrategyOptions {
```

Comment:
This feels weird to me. It's a protocol masquerading as a concrete type almost. Looking through the implementation below I'm conflicted because this design choice forces a few other choices, f.ex not instantiating an encoder/decoder in place, which I think connotes a hard reset of state in its previous form. 

Were there alternatives considered before landing on this approach? Maybe something like `BsonEncoder(fromClientOptions:)` or embedding the `CodingStrategyOptions` in the `ClientOptions`? Maybe even that `MongoDatabase` and `MongoCollection` take type-specific `options` that also have these fields, and e.g. `MongoDatabaseOptions(fromClientOptions:)`

Reply:
Doing `BSONEncoder(fromClientOptions:)` would also necessitate other initializers for collection, database, and generic options. Protocols help avoid that duplication nicely, I think.

We could decide to embed the `CodingStrategyOptions` struct within all the options types that conform to `CodingStrategyOptions` protocol (and remove the protocol) at the cost of a bit of verbosity.

e.g. The API would become
```swift
// inline
let client = MongoClient(..., options: ClientOptions(codingStrategyOptions: CodingStrategyOptions(dateCodingStrategy: .deferredToDate)))
// two step
let strategyOpts = CodingStrategyOptions(dateCodingStrategy: .deferredToDate)
let client = MongoClient(..., options: ClientOptions(codingStrategyOptions: strategyOpts))
```

The last approach would still end up requiring either the current design or one of the approaches you mentioned, so it is sort of orthogonal to the issue here. If this were a pure Swift driver and/or if we had a lot of options to pass from client->db->collection, it would definitely be nice to have conversion capability (a la `FindAndModifyOptionsConvertible`), but since the coding strategy stuff is the only options we're passing, I'm not sure how helpful it would be. 

Embedding the `CodingStrategyOptions` within other options types as necessary sounds fine to me if you'd like me to make that change.

Reply:
Maybe it's just a naming thing, like maybe `CodingStrategyProvider` would have made more sense to me on first read. Do you have thoughts on the only observation about keeping instances of encoders around vs instantiating as-needed? It seems to be the pattern in Swift's internal code that encoders/decoders are not long-lived objects.

Reply:
Yeah, `CodingStrategyProvider` is a much better name. If we keep the current approach, I'll definitely refactor the protocol to be named that.

>Do you have thoughts on the only observation about keeping instances of encoders around vs instantiating as-needed?

The encoders and decoders as we have implemented them (and as Apple has implemented the JSON ones) are essentially just factories for internal, private encoders and decoders. The top-level ones provide a nicer API, while the internal ones do most of the work (e.g. we have `_BSONEncoder`, they have `_JSONEncoder`). So, our top level encoders are essentially behaving as containers for the encoding strategies that we set on them--they don't have any other state AFAIK. I would say, as I understand it, storing the encoder and decoder is pretty much equivalent to storing a `CodingStrategyOptions` instance as part of a client/db/collection. 

As with the other part, I don't have strong feelings on this and would be happy to implement whichever way you think is best.

While looking at this, I noticed the `MongoClient` encoder/decoder are `internal`. So I will fix them to be `public` like the others.

Reply:
Cool, I'd like to wait to hear what @kmahar has to say about this.

Reply:
I think `CodingStrategyProvider` is a good name 👍 

I don't see any issue with reusing the encoders/decoders that have been customized with our desired options across operations. each call to `BSONEncoder.encode` will create a new `_BSONEncoder` from scratch. while I agree we often see something like `JSONEncoder().encode(myThing)`, those are cases where the default configuration is being used. 

one thing to keep in mind for the future is that if we add `NumberEncodingStrategy`s they may possibly conflict with how options should be encoded and sent to libmongoc. 
for example, perhaps you would specify via such strategies that `Int64`s should be encoded as `Int32`s if they are small enough. but if libmongoc expects an option as an `Int64` we need to ignore it. 
so possibly we'll need to use separate encoders for options.

Reply:
> While looking at this, I noticed the MongoClient encoder/decoder are internal. So I will fix them to be public like the others.
what's reasoning for having them be public? so you can inspect what the options are?

Reply:
>what's reasoning for having them be public? so you can inspect what the options are?

Yeah, and also so that a user can encode/decode stuff in the same manner they would when inserting into/reading from a given collection without actually reaching out to the db. Like if they read/write bson to/from another source (e.g. disk).